### PR TITLE
[shardformer] support sharded checkpoint IO for models of HybridParallelPlugin 

### PR DESCRIPTION
**Pull request by Fridge003**
## 📌 Checklist before creating the PR

- [x] I have created an issue for this PR for traceability
- [x] The title follows the standard format: `[doc/gemini/tensor/...]: A concise description`
- [x] I have added relevant tags if possible for us to better distinguish different PRs

## 🚨 Issue number

> Link this PR to your issue with words like fixed to automatically close the linked issue upon merge
>
> e.g. `fixed #1234`, `closed #1234`, `resolved #1234`

#4477

## 📝 What does this PR do?

> Summarize your work here.
> if you have any plots/diagrams/screenshots/tables, please attach them here.

- Support sharded saving for models of HybridParallelPlugin
- Support sharded loading for models of HybridParallelPlugin
- Add relevant tests

## 💥 Checklist before requesting a review

- [x] I have linked my PR to an issue ([instruction](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue))
- [x] My issue clearly describes the problem/feature/proposal, with diagrams/charts/table/code if possible
- [x] I have performed a self-review of my code
- [x] I have added thorough tests.
- [x] I have added docstrings for all the functions/methods I implemented

## ⭐️ Do you enjoy contributing to Colossal-AI?

- [x] 🌝 Yes, I do.
- [ ] 🌚 No, I don't.

Tell us more if you don't enjoy contributing to Colossal-AI.

### CODE REVIEW

File: colossalai/checkpoint_io/hybrid_parallel_checkpoint_io.py [hunk 1]
```diff
@@ -0,0 +1,339 @@
...
+
+    def save_lr_scheduler(self, lr_scheduler: LRScheduler, checkpoint: str):
+        """
+        Save lr scheduler to checkpoint but only on master process.
+        """
+        if self.coordinator.is_master():
+            super().save_lr_scheduler(lr_scheduler, checkpoint)
+
+
+class _StateDictSharder:
```
Comment (ver217) [inline colossalai/checkpoint_io/hybrid_parallel_checkpoint_io.py:317]: Is this different from https://github.com/hpcaitech/ColossalAI/blob/27061426f7f67739e27abdbd92f9826e450c53d2/colossalai/zero/gemini/gemini_ddp.py#L784

File: colossalai/checkpoint_io/hybrid_parallel_checkpoint_io.py [hunk 2]
```diff
@@ -0,0 +1,339 @@
...
+
+class HypridParallelCheckpointIO(GeneralCheckpointIO):
+    """
+    CheckpointIO for Hybrid Parallel Training.
+
+    Args:
+        pg_mesh (ProcessGroupMesh): Process group mesh containing information of process groups along different dimensions.
+    """
+
+    def __init__(self, pg_mesh: ProcessGroupMesh) -> None:
```
Comment (ver217) [inline colossalai/checkpoint_io/hybrid_parallel_checkpoint_io.py:53]: Can this class receive three process groups? This is easy to extend if `ProcessGroupMesh` changes.

File: colossalai/checkpoint_io/hybrid_parallel_checkpoint_io.py [hunk 3]
```diff
@@ -0,0 +1,339 @@
...
+        state_dict_sharder = _StateDictSharder(size_per_shard)
+
+        # Save parameters.
+        for name, param in model.named_parameters():
+            if param is None:
+                continue
+            # Gather tensor pieces when using tensor parallel.
+            param_ = param if keep_vars else param.detach()
+            if is_distributed_tensor(param_):
+                param_ = to_global(param_)
```
Comment (ver217) [inline colossalai/checkpoint_io/hybrid_parallel_checkpoint_io.py:79]: It seems that this part is similar to `ParallelModule`. Can you extract this part to a common util function?

File: colossalai/checkpoint_io/hybrid_parallel_checkpoint_io.py (hunk 1 above)
Comment (Fridge003) [reply colossalai/checkpoint_io/hybrid_parallel_checkpoint_io.py]: StateDictSharder is moved to utils.py in checkpointio

File: colossalai/checkpoint_io/hybrid_parallel_checkpoint_io.py (hunk 2 above)
Comment (Fridge003) [reply colossalai/checkpoint_io/hybrid_parallel_checkpoint_io.py]: modified

File: colossalai/checkpoint_io/hybrid_parallel_checkpoint_io.py (hunk 3 above)
Comment (Fridge003) [reply colossalai/checkpoint_io/hybrid_parallel_checkpoint_io.py]: modified

File: colossalai/checkpoint_io/hybrid_parallel_checkpoint_io.py [hunk 4]
```diff
@@ -0,0 +1,318 @@
...
+    save_state_dict,
+    save_state_dict_shards,
+)
+
+try:
+    from torch.nn.modules.module import _EXTRA_STATE_KEY_SUFFIX, _IncompatibleKeys
+except ImportError:
+    _EXTRA_STATE_KEY_SUFFIX = '_extra_state'
+
+DP_AXIS, PP_AXIS, TP_AXIS = 0, 1, 2
```
Comment (ver217) [inline colossalai/checkpoint_io/hybrid_parallel_checkpoint_io.py:46]: It seems that this line can be removed

Comment (Fridge003) [reply colossalai/checkpoint_io/hybrid_parallel_checkpoint_io.py]: removed

**Review by ver217 — APPROVED** [review]

**Pull request merged**